src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticRangeVisitor.cs

```diff
@@ -485,21 +484,42 @@ private void AddSemanticRange(SyntaxNode node, int semanticKind)
             var childNodes = node.ChildNodes();
             if (childNodes.Count == 0)
             {
-                var content = node.GetContent();
-                var lines = content.Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
                 var charPosition = range.Start.Character;
-                for (var i = 0; i < lines.Length; i++)
+                var lineStartAbsoluteIndex = node.SpanStart - charPosition;
+                for (var i = range.Start.Line; i <= range.End.Line; i++)
```

Comment:
Consider changing `i` to `line` or `lineNumber` to make the logic slightly clearer below.